File: backend/server/src/services/tools/routePlanning.js
/**
 * route_planning 工具
 * 调用高德路径规划 API，规划两个位置之间的路线
 */

const axios = require('axios');
const config = require('../../config');

// 高德路径规划 API 端点
const AMAP_DIRECTION_API_V3 = 'https://restapi.amap.com/v3/direction/';
const AMAP_DIRECTION_API_V4 = 'https://restapi.amap.com/v4/direction/';

/**
 * 出行方式映射
 * 将用户友好的出行方式映射到高德 API 的类型
 */
const MODE_MAP = {
  'walking': 'walking',
  'driving': 'driving',
  'riding': 'bicycling', // 骑行在 v4 API 中是 bicycling
  'transfer': 'transit/integrated', // 公交规划端点
};

/**
 * 出行方式中文映射
 */
const MODE_CHINESE = {
  'walking': '步行',
  'driving': '驾车',
  'riding': '骑行',
  'transit': '公交',
};

/**
 * 规划两个位置之间的路线
 * @param {string} origin - 起点经纬度，格式：lng,lat
 * @param {string} destination - 终点经纬度，格式：lng,lat
 * @param {string} mode - 出行方式：walking/driving/riding/transfer
 * @param {Object} options - 额外参数
 * @param {string} options.city - 城市名称（公交规划必需）
 * @returns {Promise<Object>} 路线规划结果
 */
async function routePlanning(origin, destination, mode = 'walking', options = {}) {
  // 验证坐标格式
  if (!origin || !destination) {
    throw new Error('起点和终点坐标不能为空');
  }

  const originCoords = origin.split(',').map(c => c.trim());
  const destCoords = destination.split(',').map(c => c.trim());

  if (originCoords.length !== 2 || destCoords.length !== 2) {
    throw new Error('坐标格式错误，应为：lng,lat');
  }

  const apiMode = MODE_MAP[mode] || 'walking';

  // 判断使用 v3 还是 v4 API
  let apiUrl;
  if (mode === 'riding') {
    // 骑行使用 v4 API
    apiUrl = `${AMAP_DIRECTION_API_V4}bicycling`;
  } else if (mode === 'transfer') {
    // 公交使用 v3 API 的 integrated 端点
    apiUrl = `${AMAP_DIRECTION_API_V3}transit/integrated`;
  } else {
    // 步行、驾车使用 v3 API
    apiUrl = `${AMAP_DIRECTION_API_V3}${apiMode}`;
  }

  try {
    const params = {
      key: config.amap.webServiceKey,
      origin: origin,
      destination: destination,
      output: 'json',
    };

    // 公交规划需要 city 参数
    if (mode === 'transfer' && options.city) {
      params.city = options.city;
    }

    const response = await axios.get(apiUrl, {
      params,
      timeout: 10000, // 10秒超时
    });

    // 根据不同的 API 响应格式检查状态
    let statusOk = false;
    let errorInfo = '';

    if (mode === 'riding') {
      // v4 API 骑行
      statusOk = response.data.errcode === 0;
      errorInfo = response.data.errmsg || '';
    } else if (mode === 'transfer') {
      // 公交
      statusOk = response.data.status === '1';
      errorInfo = response.data.info || '';
    } else {
      // 步行、驾车
      statusOk = response.data.status === '1';
      errorInfo = response.data.info || '';
    }

    if (!statusOk) {
      console.error(`高德路径规划 API 返回错误: ${errorInfo}`);
      throw new Error(`路径规划失败: ${errorInfo || '未知错误'}`);
    }

    // 解析路径规划结果
    const result = parseRouteResult(response.data, mode);

    return result;
  } catch (error) {
    if (error.response) {
      // HTTP 错误
      console.error('高德 API HTTP 错误:', error.response.status);
      throw new Error(`路径规划服务暂时不可用`);
    }
    console.error('路径规划失败:', error.message);
    throw new Error(`路径规划失败: ${error.message}`);
  }
}

/**
 * 解析高德路径规划结果
 * @param {Object} data - 高德 API 响应数据
 * @param {string} mode - 出行方式
 * @returns {Object} 格式化后的路径结果
 */
function parseRouteResult(data, mode) {
  const apiMode = MODE_MAP[mode] || 'walking';

  if (mode === 'walking') {
    // 步行
    const route = data.paths?.[0];
    if (!route) {
      return {
        distance: 0,
        duration: 0,
        path: [],
        strategy: data.tag || '',
      };
    }

    return {
      distance: parseInt(route.distance) || 0, // 米
      duration: parseInt(route.duration) || 0, // 秒
      path: parsePathToCoordinates(route.steps || []),
      strategy: data.tag || '',
      transport: MODE_CHINESE[mode] || mode,
    };
  } else if (mode === 'riding') {
    // 骑行 (v4 API)
    // v4 API 返回结构: { errcode: 0, data: { paths: [...] } }
    const route = data.data?.paths?.[0];
    if (!route) {
      return {
        distance: 0,
        duration: 0,
        path: [],
        strategy: '',
      };
    }

    return {
      distance: parseInt(route.distance) || 0,
      duration: parseInt(route.duration) || 0,
      path: parsePathToCoordinates(route.steps || []),
      strategy: '',
      transport: MODE_CHINESE[mode] || mode,
    };
  } else if (mode === 'driving') {
    // 驾车
    const route = (data.routes && data.routes.route) || data.paths?.[0];
    if (!route) {
      return {
        distance: 0,
        duration: 0,
        path: [],
        strategy: '',
      };
    }

    return {
      distance: parseInt(route.distance) || 0,
      duration: parseInt(route.time) || 0, // 高德驾车是 time 字段
      path: parseDrivingPath(route.steps || []),
      strategy: data.tag || '',
      transport: MODE_CHINESE[mode] || mode,
    };
  } else if (mode === 'transfer') {
    // 公交 (v3 API integrated)
    const route = data.route;
    if (!route || !route.transits || route.transits.length === 0) {
      return {
        distance: 0,
        duration: 0,
        path: [],
        segments: [],
      };
    }

    const transit = route.transits[0];
    return {
      distance: parseInt(route.distance) || 0,
      duration: parseInt(transit.duration) || 0,
      path: [],
      segments: parseTransitSegments(transit.segments || []),
      transport: MODE_CHINESE[mode] || mode,
      cost: transit.cost || 0, // 费用
    };
  }

  return {
    distance: 0,
    duration: 0,
    path: [],
    transport: MODE_CHINESE[mode] || mode,
  };
}

/**
 * 解析步行/骑行路径步骤为坐标点
 * @param {Array} steps - 路径步骤
 * @returns {Array} 坐标点数组
 */
function parsePathToCoordinates(steps) {
  const coordinates = [];

  for (const step of steps) {
    if (step.polyline) {
      // polyline 是坐标点用;分隔的字符串
      const points = step.polyline.split(';');
      for (const point of points) {
        const [lng, lat] = point.split(',');
        if (lng && lat) {
          coordinates.push({
            lng: parseFloat(lng),
            lat: parseFloat(lat),
          });
        }
      }
    }
  }

  return coordinates;
}

/**
 * 解析驾车路径步骤
 * @param {Array} steps - 驾车步骤
 * @returns {Array} 坐标点数组
 */
function parseDrivingPath(steps) {
  const coordinates = [];

  for (const step of steps) {
    if (step.polyline) {
      const points = step.polyline.split(';');
      for (const point of points) {
        const [lng, lat] = point.split(',');
        if (lng && lat) {
          coordinates.push({
            lng: parseFloat(lng),
            lat: parseFloat(lat),
          });
        }
      }
    }
  }

  return coordinates;
}

/**
 * 解析公交路径段
 * @param {Array} segments - 公交路径段
 * @returns {Array} 段信息数组
 */
function parseTransitSegments(segments) {
  return segments.map(seg => ({
    type: seg.type || '',
    line: seg.line?.name || '',
    startStation: seg.start_name || '',
    endStation: seg.end_name || '',
    duration: parseInt(seg.time) || 0,
    distance: parseInt(seg.walking?.distance) || 0,
  }));
}

/**
 * 估算交通时间（用于快速估算）
 * @param {number} distance - 距离（米）
 * @param {string} mode - 出行方式
 * @returns {number} 估算时间（秒）
 */
function estimateDuration(distance, mode = 'walking') {
  // 步行速度约 5km/h = 1.39m/s
  // 骑行速度约 15km/h = 4.17m/s
  // 驾车速度约 30km/h = 8.33m/s（考虑交通）

  const speeds = {
    walking: 1.39,
    riding: 4.17,
    driving: 8.33,
    transfer: 6.94, // 公交考虑换乘
  };

  const speed = speeds[mode] || speeds.walking;
  return Math.round(distance / speed);
}

module.exports = {
  routePlanning,
  parseRouteResult,
  parsePathToCoordinates,
  MODE_MAP,
  MODE_CHINESE,
  estimateDuration,
};

File: backend_new/src/tools/route_planning.py
import asyncio
import hashlib
import time
import re
from typing import Optional

import requests

from src.config import get_settings

settings = get_settings()

MAX_RETRIES = 3
RETRY_DELAY = 3  # seconds

# 城市名到城市代码映射
CITY_NAME_TO_CODE = {
    "北京": "010",
    "上海": "021",
    "杭州": "0571",
    "成都": "028",
    "西安": "029",
    "重庆": "023",
}


def _sign(params: dict) -> str:
    """Generate signature for Amap API request."""
    sorted_params = sorted(params.items(), key=lambda x: x[0])
    sign_str = "".join(f"{k}{v}" for k, v in sorted_params)
    sign_str += settings.AMAP_SECURITY_CODE
    return hashlib.md5(sign_str.encode()).hexdigest()


def _is_rate_limited(data) -> bool:
    """Check if response indicates rate limiting."""
    data_str = str(data).lower()
    return any(kw in data_str for kw in ["CUQPS_HAS_EXCEEDED", "rate limit", "频率", "limit exceeded", "10012"])


def _make_request(url: str, params: dict) -> Optional[dict]:
    """Make a request with retry logic for rate limiting."""
    if settings.AMAP_SECURITY_CODE:
        params["sig"] = _sign(params)

    for attempt in range(MAX_RETRIES):
        try:
            response = requests.get(url, params=params, timeout=15)
            data = response.json()

            if _is_rate_limited(data):
                if attempt < MAX_RETRIES - 1:
                    time.sleep(RETRY_DELAY)
                    continue
                return None

            return data
        except Exception as e:
            if attempt < MAX_RETRIES - 1:
                time.sleep(RETRY_DELAY)
                continue
            return None

    return None


def _format_duration(seconds) -> str:
    """Format seconds to human readable duration."""
    try:
        seconds = int(seconds)
    except (ValueError, TypeError):
        return "未知"
    if seconds < 60:
        return f"{seconds}秒"
    minutes = seconds // 60
    if minutes < 60:
        return f"{minutes}分钟"
    hours = minutes // 60
    remaining_minutes = minutes % 60
    if remaining_minutes == 0:
        return f"{hours}小时"
    return f"{hours}小时{remaining_minutes}分钟"


def _format_distance(meters) -> str:
    """Format meters to human readable distance."""
    try:
        meters = int(meters)
    except (ValueError, TypeError):
        return "未知"
    if meters < 1000:
        return f"{meters}米"
    km = meters / 1000
    if km == int(km):
        return f"{int(km)}公里"
    return f"{km:.1f}公里"


def _get_city_code(city_name: str) -> Optional[str]:
    """Convert city name to city code for Amap API."""
    # 先尝试直接映射
    if city_name in CITY_NAME_TO_CODE:
        return CITY_NAME_TO_CODE[city_name]
    # 尝试查询城市编码
    url = "https://restapi.amap.com/v3/config/district"
    params = {
        "key": settings.AMAP_KEY,
        "keywords": city_name,
        "subdistrict": 0,
        "output": "json",
    }
    data = _make_request(url, params)
    if data and data.get("districts"):
        return data["districts"][0].get("citycode")
    return None


def _clean_address(address: str) -> str:
    """Clean address string - remove parenthetical content and special characters."""
    if not address:
        return address
    # Remove content in parentheses and after "或" or "；"
    cleaned = re.sub(r'[（(][^）)]*[)）]', '', address)
    cleaned = re.split(r'[或；;]', cleaned)[0]
    return cleaned.strip()


def geocode_address(address: str) -> Optional[dict]:
    """
    Convert address string to coordinates via Amap Geocoder API.
    Returns dict with lng, lat or None if failed.
    """
    url = "https://restapi.amap.com/v3/geocode/geo"

    cleaned = _clean_address(address)

    params = {
        "key": settings.AMAP_KEY,
        "address": cleaned,
        "output": "json",
    }

    data = _make_request(url, params)
    if not data:
        # Fallback: try original address if cleaning changed it
        if cleaned != address:
            params["address"] = address
            data = _make_request(url, params)
        if not data:
            return None

    geocodes = data.get("geocodes", [])
    if not geocodes:
        return None

    location = geocodes[0].get("location", "")
    if not location:
        return None

    parts = location.split(",")
    if len(parts) != 2:
        return None

    return {
        "lng": float(parts[0]),
        "lat": float(parts[1]),
    }


def _parse_polyline(path_str: str) -> str:
    """Parse polyline from Amap response to simplified string format."""
    if not path_str:
        return ""
    return path_str.replace(";", ";")


async def search_transit(
    origin: str,
    destination: str,
    departure_time: Optional[str] = None,
    city: Optional[str] = None,
) -> Optional[dict]:
    """
    Search transit route via Amap Transit API v5.
    origin/destination: "lng,lat" format
    departure_time: "HHMM" format, e.g. "1200" for 12:00
    city: city name for transit search (e.g., "北京")
    Returns structured segment data or None.
    """
    url = "https://restapi.amap.com/v5/direction/transit/integrated"

    # 获取城市代码
    city_code = city
    if city and city not in CITY_NAME_TO_CODE.values():
        code = _get_city_code(city)
        if code:
            city_code = code

    params = {
        "key": settings.AMAP_KEY,
        "origin": origin,
        "destination": destination,
        "city1": city_code or "010",
        "city2": city_code or "010",
        "strategy": 0,  # 推荐模式
        "nightflag": 0,
        "date": time.strftime("%Y%m%d"),
        "show_fields": "polyline,cost,navi",
        "output": "json",
    }

    if departure_time:
        params["time"] = departure_time

    data = _make_request(url, params)
    if not data or data.get("status") != "1":
        return None

    route = data.get("route", {})
    paths = route.get("transits", [])
    if not paths:
        return None

    # 使用第一（推荐）路线
    transit = paths[0]

    # 解析 segments
    segments = []
    for seg in transit.get("segments", []):
        walking = seg.get("walking", {})
        bus = seg.get("bus", {})
        metro = seg.get("metro", {})

        # 步行路段
        if walking.get("steps"):
            for step in walking["steps"]:
                instruction = step.get("instruction", "")
                road = step.get("road_name", "") or step.get("road", "")
                distance = step.get("distance", 0)
                segments.append({
                    "instruction": instruction,
                    "road": road if road else "步行",
                    "distance": _format_distance(distance) if distance else "0米",
                })

        # 公交/地铁路段
        if bus.get("buslines"):
            for line in bus["buslines"]:
                name = line.get("name", "")
                instruction = f"乘坐{name}"
                segments.append({
                    "instruction": instruction,
                    "road": name,
                    "distance": "0米",
                })

        # 地铁路段
        if metro.get("lines"):
            for line in metro["lines"]:
                name = line.get("name", "")
                instruction = f"乘坐地铁{name}"
                segments.append({
                    "instruction": instruction,
                    "road": name,
                    "distance": "0米",
                })

    # 构建道路摘要
    road_parts = []
    for seg in segments[:5]:
        if seg["road"] and seg["road"] not in road_parts and seg["road"] not in ["步行"]:
            road_parts.append(seg["road"])
    road_summary = " → ".join(road_parts) if road_parts else "公交路线"

    distance = transit.get("distance", 0)
    # v5 API 中 duration 在 cost 对象里
    duration = transit.get("cost", {}).get("duration", 0) or 0

    # 获取 polyline（路线坐标点串）
    # v5 公交 API 的 polyline 在 segments 中的 walking 路段
    polyline_parts = []
    for seg in transit.get("segments", []):
        walking = seg.get("walking", {})
        if walking:
            for step in walking.get("steps", []):
                polyline = step.get("polyline")
                if polyline:
                    polyline_parts.append(str(polyline))
    polyline = ";".join(polyline_parts)

    return {
        "distance": _format_distance(distance) if distance else "未知",
        "duration": _format_duration(int(duration)) if duration else "未知",
        "road_summary": road_summary,
        "polyline": polyline,
        "steps": segments,
    }


async def search_driving(
    origin: str,
    destination: str,
    waypoints: Optional[list[str]] = None,
    strategy: int = 0,
) -> Optional[dict]:
    """
    Search driving route via Amap Driving API v5.
    origin/destination: "lng,lat" format
    Returns structured segment data with polyline and steps.
    """
    url = "https://restapi.amap.com/v5/direction/driving"
    params = {
        "key": settings.AMAP_KEY,
        "origin": origin,
        "destination": destination,
        "strategy": strategy,
        "show_fields": "polyline,navi,cost",
        "output": "json",
    }

    if waypoints:
        params["waypoints"] = ";".join(waypoints)

    data = _make_request(url, params)
    if not data or data.get("status") != "1":
        return None

    route = data.get("route", {})
    paths = route.get("paths", [])
    if not paths:
        return None

    path = paths[0]
    distance = path.get("distance", 0)
    duration = path.get("duration", 0)

    # 解析 steps
    steps = []
    road_parts = []
    for step in path.get("steps", []):
        instruction = step.get("instruction", "")
        road = step.get("road_name", "") or step.get("road", "")
        step_distance = step.get("step_distance", 0) or step.get("distance", 0)

        steps.append({
            "instruction": instruction,
            "road": road if road else "",
            "distance": _format_distance(step_distance) if step_distance else "0米",
        })

        if road and road not in road_parts:
            road_parts.append(road)

    road_summary = " → ".join(road_parts[:5]) if road_parts else "驾车路线"

    return {
        "distance": _format_distance(distance) if distance else "未知",
        "duration": _format_duration(duration) if duration else "未知",
        "road_summary": road_summary,
        "polyline": _parse_polyline(path.get("polyline", "")),
        "steps": steps,
    }


async def search_walking(
    origin: str,
    destination: str,
) -> Optional[dict]:
    """
    Search walking route via Amap Walking API v5.
    origin/destination: "lng,lat" format
    Returns structured segment data with polyline and steps.
    """
    url = "https://restapi.amap.com/v5/direction/walking"
    params = {
        "key": settings.AMAP_KEY,
        "origin": origin,
        "destination": destination,
        "show_fields": "polyline,navi",
        "output": "json",
    }

    data = _make_request(url, params)
    if not data or data.get("status") != "1":
        return None

    route = data.get("route", {})
    paths = route.get("paths", [])
    if not paths:
        return None

    path = paths[0]
    distance = path.get("distance", 0)
    duration = path.get("duration", 0)

    # 解析 steps
    steps = []
    road_parts = []
    for step in path.get("steps", []):
        instruction = step.get("instruction", "")
        road = step.get("road_name", "") or step.get("road", "")
        step_distance = step.get("step_distance", 0) or step.get("distance", 0)

        steps.append({
            "instruction": instruction,
            "road": road if road else "",
            "distance": _format_distance(step_distance) if step_distance else "0米",
        })

        if road and road not in road_parts:
            road_parts.append(road)

    road_summary = " → ".join(road_parts) if road_parts else "步行路线"

    return {
        "distance": _format_distance(distance) if distance else "未知",
        "duration": _format_duration(duration) if duration else "未知",
        "road_summary": road_summary,
        "polyline": _parse_polyline(path.get("polyline", "")),
        "steps": steps,
    }


async def search_riding(
    origin: str,
    destination: str,
    strategy: int = 0,
) -> Optional[dict]:
    """
    Search riding route via Amap Riding API v5.
    origin/destination: "lng,lat" format
    strategy: 0=综合, 1=推荐路线, 2=最快路线
    Returns structured segment data with polyline and steps.
    """
    url = "https://restapi.amap.com/v5/direction/bicycling"
    params = {
        "key": settings.AMAP_KEY,
        "origin": origin,
        "destination": destination,
        "strategy": strategy,
        "show_fields": "polyline,navi",
        "output": "json",
    }

    data = _make_request(url, params)
    if not data or data.get("status") != "1":
        return None

    route = data.get("route", {})
    paths = route.get("paths", [])
    if not paths:
        return None

    path = paths[0]
    distance = path.get("distance", 0)
    duration = path.get("duration", 0)

    # 解析 steps
    steps = []
    road_parts = []
    for step in path.get("steps", []):
        instruction = step.get("instruction", "")
        road = step.get("road_name", "") or step.get("road", "")
        step_distance = step.get("step_distance", 0) or step.get("distance", 0)

        steps.append({
            "instruction": instruction,
            "road": road if road else "",
            "distance": _format_distance(step_distance) if step_distance else "0米",
        })

        if road and road not in road_parts:
            road_parts.append(road)

    road_summary = " → ".join(road_parts) if road_parts else "骑行路线"

    return {
        "distance": _format_distance(distance) if distance else "未知",
        "duration": _format_duration(duration) if duration else "未知",
        "road_summary": road_summary,
        "polyline": _parse_polyline(path.get("polyline", "")),
        "steps": steps,
    }


async def route_planning(
    origin: str,
    destination: str,
    waypoints: list[str] = None,
    strategy: int = 0,
) -> dict:
    """
    Plan a route via Amap driving API v5.
    strategy: 0=速度优先, 1=费用优先, 2=距离优先
    Returns dict with route summary and steps.
    """
    url = "https://restapi.amap.com/v5/direction/driving"
    params = {
        "key": settings.AMAP_KEY,
        "origin": origin,
        "destination": destination,
        "strategy": strategy,
        "show_fields": "cost",
        "output": "json",
    }
    if waypoints:
        params["waypoints"] = ";".join(waypoints)

    data = _make_request(url, params)
    if not data or data.get("status") != "1":
        return {}

    route = data.get("route", {})
    paths = route.get("paths", [])
    if not paths:
        return {}

    return {
        "distance": paths[0].get("distance", ""),
        "duration": paths[0].get("duration", ""),
        "strategy": strategy,
    }
